sklearn/tree/tree.py

```diff
@@ -240,8 +256,17 @@ def fit(self, X, y, sample_weight=None, check_input=True,
                              "number of samples=%d" % (len(y), n_samples))
         if self.min_samples_split <= 0:
             raise ValueError("min_samples_split must be greater than zero.")
+        if not 1 < min_samples_split <= n_samples:
+            raise ValueError("min_samples_split must in [2, n_samples]")
         if self.min_samples_leaf <= 0:
             raise ValueError("min_samples_leaf must be greater than zero.")
+        elif 0.5 <= self.min_samples_leaf < 1:
+            raise ValueError("min_samples_leaf in percentage must be smaller "
+                             "than 0.5.")
+        if not 0 < min_samples_leaf <= n_samples:
+            raise ValueError("min_samples_leaf must be in (0, n_samples], "
+                             "check both the value and type of input "
+                             "parameter")
```

Comment:
But we should make the same restriction when `min_samples_leaf` is `int`, right?